File: 0001-two-sum/0001-two-sum.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Create a dictionary to store the value and its index
        num_to_index = {}

        # Iterate through the list
        for i, num in enumerate(nums):
            # Calculate the complement
            complement = target - num

            # Check if the complement exists in the dictionary
            if complement in num_to_index:
                # Return the indices of the current number and its complement
                return [num_to_index[complement], i]
            
            # Store the current number with its index
            num_to_index[num] = i

File: 0001-two-sum/0001-two-sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
    //     vector<int> v;

    //     for(int i = 0; i < nums.size(); i++)
    //     {
    //         for(int j = i + 1; j < nums.size(); j++)
    //         {
    //             if(nums[i] + nums[j] == target)
    //             {
    //                 v.push_back(i);
    //                 v.push_back(j);
    //             }
    //         }
    //     }
    //     return v;
    // }

    map<int, int> mp;

    for(int i = 0; i < nums.size(); i++)
    {
        int need = target - nums[i];

        if(mp.find(need) != mp.end())
        {
            return {mp[need], i};
        }

        mp[nums[i]] = i;
    }

    return {-1, -1};
    }
};
